topdown/eval.go

```diff
@@ -859,7 +869,107 @@ func (e *eval) biunifyComprehension(a, b *ast.Term, b1, b2 *bindings, swap bool,
 		return e.biunifyComprehensionObject(a, b, b1, b2, iter)
 	}
 
-	return fmt.Errorf("illegal comprehension %T", a)
+	return internalErr(e.query[e.index].Location, "illegal comprehension type")
+}
+
+func (e *eval) indexComprehension(a *ast.Term) (*ast.Term, error) {
+
+	index := e.comprehensionIndex(a)
```

Comment:
nit: Having both `indexComprehension` and `comprehensionIndex` on `eval` seems less than ideal, could we distinguish the names more?

Reply:
Renamed to `indexComprehension` to `buildComprehensionCache`.